---
title: "Chicdiff Vignette"
author: "Jonathan Cairns,William Orchard, Valeriya Malysheva, Mikhail Spivakov"
date: 15th November, 2018
VignetteIndexEntry: 
output:
  BiocStyle::html_document:
    toc: true

---
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Chicdiff Vignette}
%\VignetteDepends{}
%\VignetteKeywords{Chicdiff}
%\VignettePackage{Chicdiff}
-->
```{r style, echo = FALSE, results = 'asis'}
# BiocInstaller::biocLite("BiocStyle") # if needed...
BiocStyle::markdown()
```

# Introduction

Chicdiff is a method for detecting statistically significant differential interactions in Capture HiC data. This vignette will walk you through a typical Chicdiff analysis. 

A typical Chicdiff job for two biological replicates of CHi-C data takes 30 min wall-clock time and uses 15G RAM. Time and memory usage depends on data sparcity.

------ ---------------------------------------------------------
*NOTE* A wrapper to perform this type of analysis, called *runChicdiff.R*, is provided as part of *chicdiffTools*, which is available from our [Github repository](https://github.com/RegulatoryGenomicsGroup/chicdiff/). Refer to the *chicdiffTools* README for more information.


---------------------------------------------------------------

The statistical foundations of Chicdiff are presented in a separate paper that is currently available as a preprint [ref!!!]. Briefly, Chicdiff combines moderated differential testing for count data implemented in DESeq2 with CHi-C-specific procedures for signal normalisation and p-value weighting. To increase power, Chicdiff combines reads across several fragments surrounding each significantly interacting region for each bait. Background estimates from CHiCAGO are used to construct a custom scaling matrix that is used in differential testing. The count and scaling matrices are provided as input for the DESeq2 package that tests each interaction for differences between conditions using a negative binomial model with moderated dispersion estimation. DESeq2 implements both pairwise testing between conditions, as well as likelihood ratio tests for multi-condition differences. The p-values from either of these tests are then submitted to the weighting procedure as described below. Due to high distance-dependent variations of signal-to-noise ratios and effect sizes in CHi-C data Chicdiff performs non-uniform multiple testing correction, such that weaker signals at longer distances are corrected more vigorously. We use IHW to learn the weights on the p-values obtained from a differential testing of fragment pairs in a 'training set' that is sampled from the full count data for each sample. We use interaction distance bins as covariates to assign weights to each distance bin in a manner that maximises the overall fraction of rejected null hypotheses. The distance-dependent weights learned this way are applied to the p-values in the test set, and the resulting weighted p-values are reported to the user.



--------- ------------------------------------------------------
*WARNING* The data set used in this tutorial comes from the package *ChicdiffData*. This package contains small parts (chromosome 19) of published Promoter Capture HiC data sets in human Naive CD4+ T cells and Monocytes [@Javierre2016] - note that we are only reusing raw data from this publication. Do not use any of these sample input data for purposes other than training. 

---------------------------------------------------------------


In this vignette, we use the Naive CD4+ T cells and Monocytes data [@Javierre2016]:

```{r, message=FALSE}
library(Chicdiff)
library(ChicdiffData)
```


# Workflow

## Input files required

Before you start, you will need: 

1. Two restriction map information files ("design files"):

- Restriction map file (.rmap) - a bed file containing coordinates of the restriction fragments. By default, 4 columns: chr, start, end, fragmentID.
- Bait map file (.baitmap) - a bed file containing coordinates of the baited restriction fragments, and their associated annotations. By default, 5 columns: chr, start, end, fragmentID, baitAnnotation. The regions specified in this file, including their fragmentIDs, must be an exact subset of those in the .rmap file. The baitAnnotation is a text field that is used only to annotate the output and plots.

We recommend that you put all five of these files into the same directory (that we refer to as designDir). An example of a valid design folder, for a sample of the CD4+ and Monocytes data used in this vignette, is provided in the ChicdiffData package, as follows.

```{r}
dataPath <- system.file("extdata", package="ChicdiffData")
testDesignDir <- file.path(dataPath, "designDir")
dir(testDesignDir)
```

------ ---------------------------------------------------------
*NOTE* These are same files that were used to run CHiCAGO.

---------------------------------------------------------------


2. *Peakfile*, obtained from CHiCAGO run and containing the scores for the interactions considered by CHiCAGO.

Example *peakfile* file is provided in the ChicdiffData package:

```{r}
peakFiles <- file.path(dataPath,"peakMatrix_cd4_mono_unmerged.txt")
```

3. You will also need input data files, which should be in CHiCAGO input format, *.chinput*. You can obtain *.chinput* files from aligned Capture Hi-C BAM files by running ``bam2chicago.sh``, available as part of ``chicagoTools``. (To obtain BAM files from raw fastq files, use a Hi-C alignment & QC pipeline such as [HiCUP](http://www.bioinformatics.babraham.ac.uk/projects/hicup/).

Example *.chinput* files are provided in the ChicdiffData package, as follows:

```{r}
testDataPath_CD4 <- file.path(dataPath, "CD4")
testDataPath_Mono <- file.path(dataPath, "monocytes")
dir(testDataPath_CD4)
dir(testDataPath_Mono)
```

To run Chicdff *.chinput* files have to be presented in the follwing way:

```{r}
targetChs <- list(
    CD4 = c(NCD4_22 = file.path(testDataPath_CD4, "unitTest_CD41.chinput"),
            NCD4_23 = file.path(testDataPath_CD4, "unitTest_CD42.chinput")
            ),
    
    Mono = c(Mon_2 = file.path(testDataPath_Mono, "unitTest_Mono2.chinput"),
            Mon_3 = file.path(testDataPath_Mono, "unitTest_Mono3.chinput")
            )
  )
```

------ --------------------------------------------------------
*NOTE* The names of every file should match the column name of the correpsonding data set in the *peakfile* file.

---------------------------------------------------------------

4. Finally you will need *.Rda* input data files, which should be produced in CHiCAGO run. These files contain information on interactions, including their coordinates, bait- and OtherEnd-specific data etc.

Example *.Rda* files are provided in the ChicdiffData package, as follows:

```{r}
testDataPath_rda <- system.file("data", package="ChicdiffData")
dir(testDataPath_rda)
```

To run Chicdff *.Rda* files have to be presented in the follwing way:

```{r}
targetRDSorRDAs <- list(
  CD4 = c(NCD4_22 = file.path(testDataPath_rda, "unitTest_CD41.RDa"),
          NCD4_23 = file.path(testDataPath_rda, "unitTest_CD42.RDa")
          ),
  
  Mono = c(Mon_2 = file.path(testDataPath_rda, "unitTest_Mono2.RDa"),
           Mon_3 = file.path(testDataPath_rda, "unitTest_Mono3.RDa")
  )
)
```

------ --------------------------------------------------------
*NOTE* As for *targetChs*, the names of every file should match the column name of the correpsonding data set in the *peakfile* file.

---------------------------------------------------------------

## Example workflow

We run Chicdiff on the test data as follows. 
First, we set up chicdiff experiment settings with ``setchicExperiment()``:


```{r, message=FALSE, warning = FALSE}
defchic.settings <- setchicExperiment(designDir = testDesignDir, targetRDSorRDAs = targetRDSorRDAs, targetChs = targetChs, peakfiles = peakFiles)
```

OPTIONAL: You can modify settings directly in the command line. For example, to set up the running mode to parallel run the following line:

```{r, message=FALSE, warning = FALSE}
defchic.settings <- setchicExperiment(designDir = testDesignDir, targetRDSorRDAs = targetRDSorRDAs, targetChs = targetChs, peakfiles = peakFiles, settings = list(parallel=TRUE))
```

Alternatively, you can provide your prefered parameters in a *settingsFile* file:

```{r, message=FALSE, warning = FALSE}
settingsFile = file.path(dataPath,"SettingsFile.txt")
defchic.settings <- setchicExperiment(designDir = testDesignDir, targetRDSorRDAs = targetRDSorRDAs, targetChs = targetChs, peakfiles = peakFiles, settingsFile = settingsFile)
```

*Omit this step unless you know which settings you wish to alter*.

------ ---------------------------------------------------------
*NOTE* Settings override settings from settingsFile. Both override default settings.

---------------------------------------------------------------

Finally, we run the pipeline with ``chicdiffPipeline()``:

```{r, eval=FALSE}
output <- chicdiffPipeline(defchic.settings)
```

## Output diagnostic plots 

``chicdiffPipeline()`` produces a number of plots, that are saved in the working directory.

The plots produced by Chicdiff for the chr19 datasets are provided in the *ChicdiffData* package.
Note that the diagnostic plots are computed using only a small fraction of the full CD4 and Monocyte datasets. In real-world, genome-wide datasets the trends described below should be more pronounced. 

### Decision boundary
```{r}
resultsPath <- file.path(dataPath, "CD4_Mono_results")

IHWdecisionBoundaryPlot <- png::readPNG(file.path(resultsPath,"IHWdecisionBoundaryPlot.png"))
grid::grid.raster(IHWdecisionBoundaryPlot)
```
 
This plot shows the implied decision boundaries for the unweighted p-values, as a function of the covariate. It shows whether the covariate (distance) in the control dataset is informative of the power under the alternative.

The trend is: low p-values are enriched at low covariate values. The higher the covariate values, the higher are the pvalues. This indicates that the distance covariate is correlated with power under the alternative.

### Estimated weights
```{r}
IHWweightPlot <- png::readPNG(file.path(resultsPath,"IHWweightPlot.png"))
grid::grid.raster(IHWweightPlot)
```

We see that the general trend is driven by the covariate (stratum) (distance in our case). As expected, the weight functions calculated on different random subsets (folds) of the data behave similarly. For the chr19 data at hand, interactions acros longer distances get penalized and are assigned with a lower weight, while interactions over shorter distances get prioritized.

# Vizualizing differential interactions

The plotdiffBaits() function can be used to plot the raw read counts versus linear distance from bait, mirrored for compared conditions. Significant Interactions are represented as intervals, labelled in a different colour, according to the attributed weighted pvalue.

```{r}
IHWweightPlot <- png::readPNG(file.path(resultsPath,"diffbaitPlot.png"))
grid::grid.raster(IHWweightPlot)
```

By default Chicdiff produces a plot, showing four random baits from the top 100 baits, containing interactions with the lowest weighted pvalue. Interactions within 1 Mb from bait are shown, with those passing the threshold of 5 colored in red and those passing the more lenient threshold of 3 shown in blue. Those interactions that are passing the threshold of 5 go through the Chicdiff pipeline. To gain the statistical power the otherEnd under evaluation will be expanded to an interval, which will be compared between conditions. Thus the weighted pvalue correpsonds not to an individual interaction but to an interaction of bait with an interval and is the result of several interactions withing that interval.

Using the final Chicdiff output file, countput and the baitmap as inputs to *plotdiffBaits()* function the user can modify the number of baits to display or provide a set of defind baits to plot, as well as change the display parameters, such as color of interactions, the distance around the bait to display and others (see documentation of *plotdiffBaits*).

```{r, warning=FALSE}
mini_output <- readRDS(file.path(resultsPath, "06Weighted.Rds"))
mini_countput <- readRDS(file.path(resultsPath, "countput.Rds"))
bmap <- fread(file.path(testDesignDir, "chr19_GRCh37_HindIII.baitmap"))

baits <- c(331172, 320562, 327700)
plotdiffBaits(output = mini_output, countput = mini_countput, baitmapfile = bmap, baits = baits)
```

# Association of the interaction interval with the *seed* otherEnd 

It is important to keep in mind that weighted_pvalue produced by Chicdiff correpsonds not to an individual interaction but to an interaction of bait with an interval and is the result of several interactions withing that interval.

To have an idea of fragments which interaction with the bait of interest contributed to the weighted pvalue users can user *getCandidateInteractions()* function.

```{r}
output = file.path(resultsPath, "06Weighted.Rds")
peakFiles = file.path(dataPath, "peakMatrix_cd4_mono_unmerged.txt")

outpeak <- getCandidateInteractions(output = output, countput = countput, pvcut = 0.05)

head(outpeak)
```

